### DISCUSSION TITLE
classes/class_object

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#293] [User-contributed notes]
# classes/class_object

Inherited By: AudioServer, CameraServer, ClassDB, DisplayServer, EditorFileSystemDirectory, EditorInterface, EditorPaths, EditorSelection, EditorUndoRedoManager, EditorVCSInterface, Engine, EngineD...

https://docs.godotengine.org/en/stable/classes/class_object.html

**Comment by MartinNr5** [comment]
There is as far as I can tell no example that shows how to use Connect() with my own parameters *and* flags. In fact, I don't think there's an example that shows how to use flags at all.

**Reply by willnationsdev** [reply]
@MartinNr5
Put this together to help: here's an advanced, working example (see the `_reset` method) using Godot 4.3 stable.

```gdscript
# log.gd, an autoload singleton named "Log".
extends Node
## An example logging singleton. Very verbose considering warnings/errors
## already include this info, but it illustrates the point.

## This creates a "global" contact point for any 3rd-party API that wants to
## know when a log has occurred (e.g. to also log elsewhere w/o editing this code).
signal log_requested(message: String, level: int, category: String)

enum Level {
	NONE,
	VERBOSE,
	DEBUG,
	INFORMATION,
	WARNING,
	ERROR,
	FATAL
}

var targets: Array[String] = ["godot", "file"]:
	get: return targets
	set(value):
		targets = value
		_reset()

var default_level := Level.INFORMATION

## Set to non-empty string if you want a console log any time this class
## attempts to print something.
var meta_msg := ""


func _ready() -> void:
	_reset()


## Enqueues multiple async one-off logging jobs for end of frame, one for each target.
##
## - We need CONNECT_DEFERRED to wait till end of frame.
## - We need CONNECT_REFERENCE_COUNTED because we bind the same function on the same
##     object to the same signal, even if we buffer different extra arguments via `bind()`.
##
## Note: because we bind `meta_msg` using a Callable, you can call `defer_log`, immediately
##     change the `meta_msg` property, and see the updated value when the deferred log triggers.
func _reset() -> void:
	for conn in log_requested.get_connections():
		(conn.signal as Signal).disconnect(conn.callable)
	var flags = CONNECT_DEFERRED | CONNECT_REFERENCE_COUNTED
	var get_meta_msg := func(): return meta_msg
	for target in targets:
		log_requested.connect(_log_target.bind(get_meta_msg).bind(target), flags)


## More user-friendly abstracted API for using the signal.
func defer_log(message: String, level: int, category: String) -> void:
	log_requested.emit(message, level, category)


## Logs a message to a specific target immediately.
## Note: if you add type hints to `target` or `meta`, then the `.bind(...)` calls we make
##     in `_reset()` will produce an error, even if the type matches. I think it's because
##     the Godot API loses the type information in the binding process & so thinks
##     that we are passing an untyped value into a typed parameter.
func _log_target(message: String, level: int, category: String, target, meta) -> void:
	# For fun, I included this to also demonstrate `bind` arguments for Callables.
	if meta:
		var meta_val := ""
		match typeof(meta):
			TYPE_STRING: meta_val = meta
			TYPE_CALLABLE when meta.is_valid(): meta_val = meta.call()
		if meta_val:
			print("Log._log_target[meta]: %s" % meta_val)

	# WARNING: Actual logging should use a more precise method than this.
	var datetime_utc := Time.get_datetime_string_from_system(true)

	var actual := "%s [%s] %s: %s" % [datetime_utc, level2str(level), category, message]
	if target == "godot":
		match level:
			Level.FATAL, Level.ERROR:
				push_error(actual)
			Level.WARNING:
				push_warning(actual)
			_:
				print(actual)
	elif target == "file":
		pass # logs to a rolling file or something.


func level2str(level: int) -> String:
	match level:
		Level.NONE: return level2str(default_level)
		Level.VERBOSE: return "VRB"
		Level.DEBUG: return "DBG"
		Level.INFORMATION: return "INF"
		Level.WARNING: return "WRN"
		Level.ERROR: return "ERR"
		Level.FATAL: return "FTL"
		_: return "???"
```

Example Usage:

```gdscript
# my_node.gd
extends Node


func _ready() -> void:
	var category = get_script().resource_path.get_file().get_basename()
	
	Log.defer_log("A deferred message with formatting that prints at end of frame", \
		Log.Level.INFORMATION, \
		category)

	Log.defer_log("Another deferred message!", \
		Logs.Level.INFORMATION, \
		category)

	print("This should print first, without formatting")
```

Prints (when I tested it):

```
This should print first, without formatting
2024-12-30T22:24:10 [INF] my_node: A deferred message with formatting that prints at end of frame
2024-12-30T22:24:10 [INF] my_node: Another deferred message!
```

**Comment by RobynTheDevil** [comment]
Requesting documentation and examples for overloading/shadowing new() on Object.

Observations:
Currently cannot call super() in overridden new
```gdscript
extends Node

class Foo extends Object:
    static func new():
    	var this = super()
    	return this

func _ready():
	var __ = Foo.new()


> Function "new()" not found in base Object.
```

Workaround is to reference Object directly
```gdscript
...
        var this = Object.new()
```

Casting up appears implicit
```gdscript
...
func _ready():
	var __ = Foo.new()
    print(str(__ is Foo))

> true
```

**Reply by RobynTheDevil** [reply]
The previous is incorrect, new is not called on the extending class at all.

**Reply by willnationsdev** [reply]
If I recall correctly, GDScript uses an `_init()` function that the engine will call during a constructor, if it is defined. Any parameters you add to that method will become required parameters when instantiating the class. And I think the Godot Editor itself will use empty constructors when making instances of your script, so you will also often want it to have default values for whatever parameters are there (though, maybe that is only needed for tool scripts?).

**Reply by Calinou** [reply]
- See https://github.com/godotengine/godot-proposals/issues/1513.

**Comment by manuq** [comment]
I notice that `set_process(true)` has to be called always after setting the object script. This is regardless of the process mode of the object. Should this be documented in [set_script()](https://docs.godotengine.org/en/4.3/classes/class_object.html#class-object-method-set-script)? Here is an example (isolated from the [BlockCoding plugin](https://github.com/endlessm/godot-block-coding/blob/f3eea034faa0580a2be0c78ece42c68f4f78f45b/addons/block_code/block_code_node/block_code.gd#L69-L70)).

```GDScript
extends Node2D

@onready var icon: Sprite2D = %Icon

var source_code_a: String = """
extends Sprite2D

func _process(delta: float) -> void:
	rotation_degrees -= 1
"""

var source_code_b: String = """
extends Sprite2D

func _process(delta: float) -> void:
	rotation_degrees += 3
"""

# Called when the node enters the scene tree for the first time.
func _ready() -> void:
	_override_icon_script(source_code_a)


func _override_icon_script(source_code: String):
	var script := GDScript.new()
	script.set_source_code(source_code)
	script.reload()
	icon.set_script(script)
	icon.set_process(true)


func _process(delta: float) -> void:
	if Input.is_action_just_released("ui_left"):
		_override_icon_script(source_code_a)
	elif Input.is_action_just_released("ui_right"):
		_override_icon_script(source_code_b)

```

**Comment by Wilker-uwu** [comment]
what is `*_meta` about? is this in any way related to `get_property_list`?

**Reply by willnationsdev** [reply]
The `get_meta`, `set_meta`, and `has_meta` methods are for editing/querying metadata on Godot objects. It is like a hidden `Dictionary<string, Variant>` that stores transient, dynamic, and ad-hoc data. Unlike properties, this data is not formally defined (it is not in the `ClassDB` nor any `Script`), and, because it is both stored *and defined* directly on an individual instance, there is no hierarchical/inheritance key lookup mechanism like there is with the virtual `_get` and `_set` methods for properties.

I've most often observed it being used to add boolean flags for caching or compatibility-related checks for some algorithm; also for adding group/tag like functionality to non-Node objects. And lastly, when a script needed to create a static singleton and chose to put an object instance inside a meta value on the script itself.

If you're familiar with any web programming, it's like the `HttpContext.Items` property. Or kind of like environment variables for systems programming. Just a shared, low-level API for key-value data that can be used as the need arises.

**Comment by OeilDeLance** [comment]
I could not override GodotObject._Init does not exist in C#, it would seem it does not exist.
Shouldn't that be reflected in the doc ?
Is there an alternative to it ?
I am trying to have LocalToScene be true by default on a particular Resource of mine.

**Reply by GabCoolDude** [reply]
You can just use C# constructors. It's probably documented somewhere, but to find it is another story.

**Comment by wrstldev** [comment]
If get_class() and is_class() ignore my class names, how am I supposed to find my custom class nodes when scanning the node tree? Some clarifying statements would be helpful here because I have been trying to find this out for 3 or 4 hours now. All the threads I've found online are just pages long arguments with no one posting the answer.

----

So I finally found out.

We can reword the last sentence here to say:

Note: This method ignores class_name declarations in the object's script. Instead use the 'is' keyword to find out if a node is a custom class type. Example if a is b:

The lack of clarity here is a huge time waster anyone trying to figure this out.

**Comment by Varonex** [comment]
I am a bit confused why some methods return `Variant` instead of more precise types. Example being `get_script` is said to return a `Variant`, and I don't really know why the return type doesn't ensure a `Script` instead. Could someone cast a light on this?

**Comment by moonvent** [comment]
CONNECT_PERSIST doesn't work if you setup signal with lambda function, just add this to doc

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot/pull/112487

**Comment by GersonFeDutra** [comment]
Should `notify_property_list_changed` always be used instead of emitting the signal `property_list_changed` directly?

**Reply by GabCoolDude** [reply]
Doesn't matter, they both do the same thing internally, use whichever one you prefer.

**Reply by GersonFeDutra** [reply]
Why the redundancy then?

**Reply by raijoel699-sudo** [reply]
this part seems repeated maybe we can remove one

**Reply by GabCoolDude** [reply]
Both have been there since 4.0 and `notify_property_list_changed` was introduced to replace another method. The only one you _could_ remove is `notify_property_list_changed`, but it'd be a breaking change and compared to `property_list_changed.emit()`, it's more verbose, which IMHO is better for beginners trying to use tool scripts.

**Comment by Varonex** [comment]
I am doing a sortof reflection exercise on external scripts (not inbuilt). I have found two sets of quirks that the documentation may need to indicate.
### get_property_list
- The first entry specifies the filename of the script. It is not a defined property.
- `@export_group` are also contained within the property list.

### get_method_list
- Properties defined with a getter/setter syntax implies that inbuilt methods prefixed with `@` are automatically generated, and listed. E.g:
```gdscript
var my_property: int = 1:
	get: return my_property
	set(v): my_property = v
```
It will generate a method `@my_property_get()` for the `get:` block, and `@my_property_set(v)` for the `set(v):` block. It is normal behavior of course, as we can also explicit this behavior by writing those properties and doing `get = my_property_get` and `set = my_property_set` directly.
In general, `@*_get` & `@*_set` are generated for the property `*` depending on how it is defined. Those methods are therefore listed.

***

## Code

This is the prototype code. I commented out a few lines with `is_valid_ascii_identifier()` calls on variable names to exclude most of those edge cases :

<details>
<summary>GDscript code sample</summary>

```gdscript
class_name Reflection extends Node

# ENGINE

func _ready() -> void:
	print("Scan of all customized classes")
	print("Internal classes may be scanned with ClassDB. Customized ones can be scanned with ProjectSettings & the Script class.")
	print("BEGINNING OF SCAN ===================\n\n")
	
	# Scans custom classes only
	var classes: Array[Dictionary] = ProjectSettings.get_global_class_list()

	# {base, class, icon, language, path}
	for class_info in classes:
		_scan_class(ClassDefinition.new(class_info))
		print()
	
	print("\n\nEND OF SCAN ====================")

# METHODS

func _scan_class(class_info: ClassDefinition) -> void:
	print("Found class \"%s\" (extends \"%s\") written in %s at \"%s\""
			% [class_info.class_ident, class_info.base, class_info.language, class_info.path]
	)
	
	# Script loading
	var script: Script = load(class_info.path)
		
	# Constants
	if not script.get_script_constant_map().is_empty():
	
		print("Loading constants ...")
		var constants: Dictionary = script.get_script_constant_map()
		for constant: String in constants:
			_scan_constant(ConstantDefinition.new(constant, constants[constant]))
		print()
	
	# Properties
	if script.get_script_property_list().size() > 1:
	
		print("Loading properties ...")
		for property: Dictionary in script.get_script_property_list():
			# Avoids invalid idents (which can happen. An array entry contains the gdscript ref)
			#if not property.name.is_valid_ascii_identifier():
			#	continue

			_scan_property(PropertyDefinition.new(property))
		print()
		
	# Signals
	if not script.get_script_signal_list().is_empty():
	
		print("Loading signals ...")
		for sig: Dictionary in script.get_script_signal_list():
			_scan_method(MethodDefinition.new(sig), true)
		print()
	
	# Methods
	if not script.get_script_method_list().is_empty():
	
		print("Loading methods ...")
		for method: Dictionary in script.get_script_method_list():
			# Avoids generated methods (especially get/set-generated methods) and invalid method names
			#if not method.name.is_valid_ascii_identifier():
			#	continue
			
			_scan_method(MethodDefinition.new(method))
		print()

func _scan_constant(constant_info: ConstantDefinition) -> void:
	print("Found constant \"%s\" with value \"%s\""
			% [constant_info.constant_ident, str(constant_info.value)]
	)

func _scan_property(property_info: PropertyDefinition, is_param: bool = false) -> void:
	print("%s \"%s\" of type \"%s\"%s"
			% [
				"Found property" if not is_param else "- with param",
				property_info.property_ident,
				type_string(property_info.type),
				(" (extends \"%s\")" % property_info.property_class) if property_info.property_class else ""
			]
	)

func _scan_method(method_info: MethodDefinition, is_signal: bool = false) -> void:
	var n_args: int = method_info.args.size()
	print("Found %s \"%s\" with %d parameter%s"
			% [
				"method" if not is_signal else "signal",
				method_info.method_ident,
				n_args,
				"s" if n_args != 1 else ""
			]
	)
	
	for arg: Dictionary in method_info.args:
		_scan_property(PropertyDefinition.new(arg), true)

# INNER

class ClassDefinition:
	# VARIABLES
	
	var base: String
	var class_ident: String
	var icon: String
	var language: String
	var path: String
	
	# CONSTRUCTOR
	
	func _init(info: Dictionary) -> void:
		base = info.base
		class_ident = info.class
		icon = info.icon
		language = info.language
		path = info.path

class ConstantDefinition:
	# VARIABLES
	
	var constant_ident: String
	var value: Variant
	
	# CONSTRUCTOR
	
	func _init(ident: String, val: Variant) -> void:
		constant_ident = ident
		value = val

class PropertyDefinition:
	# VARIABLES
	
	var property_ident: String
	var property_class: StringName
	var type: int
	var hint: PropertyHint
	var hint_string: String
	var usage: PropertyUsageFlags
	
	# CONSTRUCTOR
	
	func _init(info: Dictionary) -> void:
		property_ident = info.name
		property_class = info.class_name
		type = info.type
		hint = info.hint
		hint_string = info.hint_string
		usage = info.usage

class MethodDefinition:
	# VARIABLES
	
	var method_ident: String
	var args: Array
	var default_args: Array
	var flags: MethodFlags = METHOD_FLAGS_DEFAULT
	var id: int
	var return_type: Dictionary
	
	# CONSTRUCTOR
	
	func _init(info: Dictionary) -> void:
		method_ident = info.name
		args = info.args
		default_args = info.default_args
		flags = info.flags
		id = info.id
		return_type = info.return
```
</details>
